src/node/interfaces.cpp

```diff
@@ -942,9 +943,94 @@ class BlockTemplateImpl : public BlockTemplate
         return chainman().ProcessNewBlock(block_ptr, /*force_processing=*/true, /*min_pow_checked=*/true, /*new_block=*/nullptr);
     }
 
+    std::unique_ptr<BlockTemplate> waitNext(BlockWaitOptions options) override
+    {
+        // Delay calculating the current template fees, just in case a new block
+        // comes in before the next tick.
+        CAmount current_fees = -1;
+
+        // Alternate waiting for a new tip and checking if fees have risen.
+        // The latter check is expensive so we only run it once per second.
+        auto now{NodeClock::now()};
+        const auto deadline = now + options.timeout;
+        const MillisecondsDouble tick{1000};
+
+        bool tip_changed{false};
+        // Helper to check if the tip has changed, and also update tip_changed.
+        auto check_tip_changed = [this, &tip_changed]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+            AssertLockHeld(notifications().m_tip_block_mutex);
+            if (tip_changed) return true;
+            const auto tip_block{notifications().TipBlock()};
+            // This is an instance method on BlockTemplate and no template
+            // could have been generated before a tip exists.
+            Assume(tip_block);
+            tip_changed = tip_block != m_block_template->block.hashPrevBlock;
```

Comment:
In commit "Add waitNext() to BlockTemplate interface" (7f6f58bad8e56c9a25ffbc0d105bed37e9958fef)

Seem like it would be more correct (and more compact) to write this as:

tip_changed = Assume(tip_block) && tip_block != m_block_template->block.hashPrevBlock;

Since if tip_block is null the right thing to do is just wait for it to be non-null, not set tip_changed to true and try to return a new block template below.

Reply:
> Assume(tip_block) ... if tip_block is null the right thing to do is just wait ...

Hmm, if `tip_block` is null then `Assume()` will not wait but stop the program?

Reply:
Taken

`Assume(x)` is `x` in production, so in the event of a bug `tip_changed` will be false. So the function will just wait for the deadline and return nothing.